akka-stream/src/main/scala/akka/stream/io/Framing.scala

```diff
@@ -0,0 +1,173 @@
+      var count = length
+      var decoded = 0
+      while (count > 0) {
+        decoded >>>= 8
+        decoded += (bs.next().toInt & 0xFF) << highestNibble
+        count -= 1
+      }
+      decoded & Mask
+    }
+  }
+
+  private class DelimiterFramingStage(separatorBytes: ByteString, maximumLineBytes: Int) extends PushPullStage[ByteString, ByteString] {
+    private val firstSeparatorByte = separatorBytes.head
+    private var buffer = ByteString.empty
+    private var nextPossibleMatch = 0
+    private var finishing = false
+
+    override def onPush(chunk: ByteString, ctx: Context[ByteString]): SyncDirective = {
+      buffer ++= chunk
+      doParse(ctx)
+    }
+
+    override def onPull(ctx: Context[ByteString]): SyncDirective = {
+      doParse(ctx)
+    }
+
+    override def onUpstreamFinish(ctx: Context[ByteString]): TerminationDirective = {
+      if (buffer.nonEmpty) ctx.absorbTermination()
+      else ctx.finish()
+    }
+
+    private def tryPull(ctx: Context[ByteString]): SyncDirective = {
+      if (ctx.isFinishing) ctx.fail(new IllegalStateException(
+        "Stream finished but there was a truncated final frame in the buffer"))
+      else ctx.pull()
+    }
+
+    @tailrec
+    private def doParse(ctx: Context[ByteString]): SyncDirective = {
+      val possibleMatchPos = buffer.indexOf(firstSeparatorByte, from = nextPossibleMatch)
+      if (possibleMatchPos > maximumLineBytes)
+        ctx.fail(new IllegalStateException(s"Read ${buffer.size} bytes " +
+          s"which is more than $maximumLineBytes without seeing a line terminator"))
+      else {
+        if (possibleMatchPos == -1) {
+          // No matching character, we need to accumulate more bytes into the buffer
+          nextPossibleMatch = buffer.size
+          tryPull(ctx)
+        } else if (possibleMatchPos + separatorBytes.size > buffer.size) {
+          // We have found a possible match (we found the first character of the terminator
+          // sequence) but we don't have yet enough bytes. We remember the position to
+          // retry from next time.
+          nextPossibleMatch = possibleMatchPos
+          tryPull(ctx)
+        } else {
+          if (buffer.slice(possibleMatchPos, possibleMatchPos + separatorBytes.size)
+            == separatorBytes) {
+            // Found a match
+            val parsedFrame = buffer.slice(0, possibleMatchPos).compact
+            buffer = buffer.drop(possibleMatchPos + separatorBytes.size)
+            nextPossibleMatch -= possibleMatchPos + separatorBytes.size
+            if (ctx.isFinishing && buffer.isEmpty) ctx.pushAndFinish(parsedFrame)
+            else ctx.push(parsedFrame)
+          } else {
+            nextPossibleMatch += 1
+            doParse(ctx)
+          }
+        }
+      }
+    }
+
+  }
+
+  class LengthFieldFramingStage(
+    val lengthFieldLength: Int,
+    val lengthFieldOffset: Int,
+    val byteOrder: ByteOrder) extends PushPullStage[ByteString, ByteString] {
```

Comment:
I would rather take a function Int => Int so the user can tune the length field decoding as much as he wants.